### PR TITLE
Allowing longer shot gun reads for kmers analysis

### LINKED ISSUE #123: qProfiler require new feature to handle Long reads (> 265mers )
Our current qprofiler can handle long reads up to 265 mers. However, any reads above this size it will throw exceptions, such as java.lang.IndexOutOfBoundsException

One of the main reason is we store kmers scores in one big array. When kmers size is 6 mers the possible 6mers string patterns will be 15625. If a read has 400 mers, the array size will be 400 * 15625.  The length of new long reads are various, it can be a few hundred to half million, even big. Obviously one dimension big array can't cope with such a big size. 

The new feature to handle long reads can be multi-dimension array to store the kmers score. But we not sure how long the long reads can be. For example, there are only a couple long reads in a fastq file, the length is almost chromosome size. then most element value in this new multi-dimension array will be zero or one. 

Another way is we just skip kmers part for qprofiler run when the input is a long reads file.  But we should also investigate whether other functions of qprofiler can support long reads.

### PR DESCRIPTION
**Pull request by ChristinaXu2017**
# Description

qProfiler2 used the default hardcoded value (MAX_CAPACITY = 2048 * 2048) as the maximum value for QCMGArray size. It works fine for reads less than 267 bases. Recently the shotgun sequence is getting longer and longer, we got three BAMs which contains read length reached 400 bases. So we have to increase the acceptable read length for Kmers calculation. Here we hardcoded the maximum read length to 1000 base, so the max_capacity for QCMGArray is 1000 * (5^6) = 15,625,000. 

Regarding issue#123, this tool is designed for shotgun sequence, which length is less than 500 as we saw. It does not need to be bigger than 1000. We believe 1000 is conservative enough!

Fixes #123 

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

I added a related unit test to test read sequence around the boundary value (mers=1000)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: qprofiler2/src/org/qcmg/qprofiler2/bam/BamSummaryReport2.java [hunk 1]
```diff
@@ -314,7 +314,7 @@ private void createPOS(Element parent){
 	 * Parse a SAMRecord Collate various pieces of info from the SAMRecord ready for the summariser to retrieve
 	 * @param record SAMRecord next row in file
 	 */
-	public void parseRecord( final SAMRecord record ) {
+	public void parseRecord( final SAMRecord record ) throws Exception{
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/bam/BamSummaryReport2.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/src/org/qcmg/qprofiler2/fastq/FastqSummaryReport.java [hunk 2]
```diff
@@ -107,7 +107,7 @@ public void toXml(Element parent1) {
 	 * 
 	 * @return next row in file
 	 */
-	public void parseRecord(FastqRecord record) {
+	public void parseRecord(FastqRecord record) throws Exception {
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/fastq/FastqSummaryReport.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 3]
```diff
@@ -102,7 +98,7 @@ public void getPossibleKmerStringTest()  {
 	}
 	
 	@Test
-	public void bothReversedTest() throws IOException {		
+	public void bothReversedTest() throws Exception {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 4]
```diff
@@ -319,5 +316,39 @@ private static void createTestSamFile( ) {
 					Level.WARNING, "IOException caught whilst attempting to write to SAM test file: " + input.getAbsolutePath(), e);
 		}  
 	}
-	
+		 
+	@Test
+	public void maxLengthTest() throws Exception {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 5]
```diff
@@ -265,13 +261,14 @@ else if(cycle == 4 && base.equals("AC"))
 	 * here we set the base to more the 15 base
 	 * @throws ParserConfigurationException
 	 */
-	public void bothFirstTest() throws ParserConfigurationException {
+	public void bothFirstTest() throws Exception {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 6]
```diff
@@ -214,7 +210,7 @@ public void toXmlTest() throws IOException, DOMException, ParserConfigurationExc
 	
 	
 	@Test
-	public void bothForwardTest() throws IOException {
+	public void bothForwardTest() throws Exception {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 7]
```diff
@@ -158,7 +154,7 @@ public void toXmlFastqTest() throws ParserConfigurationException {
 	}
 
 	@Test
-	public void toXmlTest() throws IOException, DOMException, ParserConfigurationException {		
+	public void toXmlTest() throws Exception {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 8]
```diff
@@ -126,7 +122,7 @@ public void bothReversedTest() throws IOException {
 	}
 	
 	@Test
-	public void toXmlFastqTest() throws ParserConfigurationException {
+	public void toXmlFastqTest() throws Exception {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java [hunk 9]
```diff
@@ -97,7 +106,11 @@ public static String producer(final int k, final String mers , final boolean inc
 		String str1 = producer( k, "", includeN );			
 		return str1.split(Constants.COMMA_STRING);
 	}	
-	public void parseKmers( byte[] readString, boolean reverse , int flagFirstOfPair){
+	public void parseKmers( byte[] readString, boolean reverse , int flagFirstOfPair) throws Exception{
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java]: are any exceptions thrown from this method? If not, please remove, otherwise be more specific in terms of the type of exception

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java [hunk 10]
```diff
@@ -97,7 +106,11 @@ public static String producer(final int k, final String mers , final boolean inc
...
+		 if( readString.length > maxCycleNo ) {
+			 throw new Exception("Can't handle large read sequence, which length is greater than " + maxCycleNo + "!");
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java]: Don't throw an Exception, use one of its subclasses.
IllegalArgumentException would be appropriate here

File: qprofiler2/src/org/qcmg/qprofiler2/bam/BamSummaryReport2.java (hunk 1 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/bam/BamSummaryReport2.java]: updated

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java (hunk 10 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java]: updated, good suggestion

File: qprofiler2/src/org/qcmg/qprofiler2/bam/BamSummaryReport2.java [hunk 11]
```diff
@@ -314,7 +314,7 @@ private void createPOS(Element parent){
 	 * Parse a SAMRecord Collate various pieces of info from the SAMRecord ready for the summariser to retrieve
 	 * @param record SAMRecord next row in file
 	 */
-	public void parseRecord( final SAMRecord record ) {
+	public void parseRecord( final SAMRecord record ) throws IllegalArgumentException{
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/bam/BamSummaryReport2.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

File: qprofiler2/src/org/qcmg/qprofiler2/fastq/FastqSummaryReport.java [hunk 12]
```diff
@@ -107,7 +107,7 @@ public void toXml(Element parent1) {
 	 * 
 	 * @return next row in file
 	 */
-	public void parseRecord(FastqRecord record) {
+	public void parseRecord(FastqRecord record) throws IllegalArgumentException {
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/fastq/FastqSummaryReport.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java [hunk 13]
```diff
@@ -97,7 +106,11 @@ public static String producer(final int k, final String mers , final boolean inc
 		String str1 = producer( k, "", includeN );			
 		return str1.split(Constants.COMMA_STRING);
 	}	
-	public void parseKmers( byte[] readString, boolean reverse , int flagFirstOfPair){
+	public void parseKmers( byte[] readString, boolean reverse , int flagFirstOfPair) throws IllegalArgumentException{
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/summarise/KmersSummary.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 14]
```diff
@@ -102,7 +100,7 @@ public void getPossibleKmerStringTest()  {
 	}
 	
 	@Test
-	public void bothReversedTest() throws IOException {		
+	public void bothReversedTest() throws IllegalArgumentException, IOException {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 15]
```diff
@@ -158,7 +156,7 @@ public void toXmlFastqTest() throws ParserConfigurationException {
 	}
 
 	@Test
-	public void toXmlTest() throws IOException, DOMException, ParserConfigurationException {		
+	public void toXmlTest() throws IllegalArgumentException, ParserConfigurationException, IOException {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 16]
```diff
@@ -214,7 +212,7 @@ public void toXmlTest() throws IOException, DOMException, ParserConfigurationExc
 	
 	
 	@Test
-	public void bothForwardTest() throws IOException {
+	public void bothForwardTest() throws IllegalArgumentException, IOException {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 17]
```diff
@@ -265,13 +263,14 @@ else if(cycle == 4 && base.equals("AC"))
 	 * here we set the base to more the 15 base
 	 * @throws ParserConfigurationException
 	 */
-	public void bothFirstTest() throws ParserConfigurationException {
+	public void bothFirstTest() throws IllegalArgumentException {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

File: qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java [hunk 18]
```diff
@@ -319,5 +318,39 @@ private static void createTestSamFile( ) {
 					Level.WARNING, "IOException caught whilst attempting to write to SAM test file: " + input.getAbsolutePath(), e);
 		}  
 	}
-	
+		 
+	@Test
+	public void maxLengthTest() throws IllegalArgumentException {
```
Comment (holmeso) [inline qprofiler2/test/org/qcmg/qprofiler2/summarise/KmersSummaryTest.java]: IllegalArgumentException is an unchecked exception and so you don't need it in the throws clause

**Comment by ChristinaXu2017** [comment]
I am sorry that the title was neglected.  done now!

**Comment by ChristinaXu2017** [comment]
yes,  it is not necessary to handle unchecked Exception. Just doubt why my eclipse didn't warn me.
